### PR TITLE
WIP: Add Bitbucket Server Support

### LINKED ISSUE #515: BitBucket Server Support
Opening a separate issue to track BitBucket server support. Better to keep #396 focussed on GitLab.

I'm about to start playing around with it now based on @orta's comment https://github.com/danger/danger-js/issues/396#issuecomment-337295471.

### PR DESCRIPTION
**Pull request by azz**
Closes #515

Still a bunch of things to do but it is starting to take shape.

TODO:

- [x] Write tests
- [x] Implement `getPullRequestDiff` - Not sure if there's a way to download the raw patch...
- [x] Format in Markdown rather than HTML
- [x] Manual test build status
- [x] Manual test comment placement, update, delete
- [ ] Documentation

### CODE REVIEW

File: source/platforms/bitbucket_server/BitBucketServerAPI.ts [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+  }
+
+  getPullRequestCommits = async (): Promise<BitBucketServerCommit[]> => {
+    const path = `${this.getPRBasePath()}/commits`
+    const res = await this.get(path)
+    return (await res.json()).values
+  }
+
+  getPullRequestDiff = async () => {
+    // TODO: possible?
```
Comment (azz) [inline source/platforms/bitbucket_server/BitBucketServerAPI.ts]: Can't seem to find an endpoint to get a raw `diff`, but there's a structured version in the REST API. Could use that but would require changing some common code.

**Comment by orta** [comment]
Wow, first of all, awesome 👍

**Comment by orta** [comment]
Might be about an hour or two before I can think properly about this

**Comment by azz** [comment]
No worries - I need to take a break so I'll pick this up again tomorrow.

File: source/platforms/bitbucket_server/BitBucketServerAPI.ts (hunk 1 above)
Comment (orta) [reply source/platforms/bitbucket_server/BitBucketServerAPI.ts]: OK, I'm reasonably sure the two underlying features that rely on this are:

* Finding out the modified/created/deleted files
* Generating metadata for a specific file (e.g. maybe a JSON diff, or showing before after etc) 

There may be two different APIs that could get that kind of information from?

File: source/platforms/BitBucketServer.ts [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+  /**
+   * Returns the response for the new comment
+   *
+   * @param {string} comment you want to post
+   * @returns {Promise<any>} JSON response of new comment
+   */
+  createComment = (comment: string) => this.api.postPRComment(comment)
+
+  // In Danger RB we support a danger_id property,
+  // this should be handled at some point
```
Comment (orta) [inline source/platforms/BitBucketServer.ts]: This is handled, I should remove the comment

File: yarn.lock [hunk 3]
```diff
@@ -1371,9 +1371,9 @@ cli-spinners@^0.1.2:
   version "0.1.2"
   resolved "https://registry.yarnpkg.com/cli-spinners/-/cli-spinners-0.1.2.tgz#bb764d88e185fb9e1e6a2a1f19772318f605e31c"
```
Comment (orta) [inline yarn.lock]: ah, I bet this is me accepting greenkeeper PRs

**Comment by azz** [comment]
Not nearly as pretty as GitHub 😢, but it works! 

image

**Comment by azz** [comment]
Since BitBucket server doesn't support HTML in markdown, I've replaced the table with a list of blockquotes, otherwise the warnings cannot be multi-line.

image

**Comment by azz** [comment]
Build Status works:

image

**Comment by azz** [comment]
https://github.com/danger/danger-js/blob/6f1371dc78ab6480480c0eb0cd992a36fb7f2d58/source/ci_source/providers/Travis.ts#L55-L57

Noticed this in all the ci_sources. But it doesn't seem to be used?

**Comment by orta** [comment]
re: 
> get supportedPlatforms() 

Yeah, so only certain CI sources will support things like GitLab/BB/BBS - for example TeamCity supports GH/GitLab/BB/BBS in Ruby: [code](https://github.com/danger/danger/blob/132c5d1a415656bee49d451b861569a6a08cfc30/lib/danger/ci_source/teamcity.rb#L95
)


So you can verify that the CI hooks up to the platform, but TBH, it might not need it, considering that the ENV vars added must mean that someone knows it will work, so maybe the abstraction isn't needed?

**Comment by orta** [comment]
Re: the comment, tricky yeah. Another option is to do some funky markdown by adding line separators between the HTML content and the markdown

```md
<table>
  <thead>
    <tr>
      <th width="5%"></th>
      <th width="95%" data-danger-table="true">Header</th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:+1:</td>
      <td>

some [markdown](/link)

</td>
    </tr>
  </tbody>
</table>
```

<table>
  <thead>
    <tr>
      <th width="5%"></th>
      <th width="95%" data-danger-table="true">Header</th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:+1:</td>
      <td>

some [markdown](/link)

</td>
    </tr>
  </tbody>
</table>

Which is the trick used in the gitlab markdown renderer for GitLab + BBS [in Ruby](https://github.com/danger/danger/blob/132c5d1a415656bee49d451b861569a6a08cfc30/lib/danger/comment_generators/gitlab.md.erb)

**Comment by azz** [comment]
It doesn't seem to support _any_ HTML at all. :(

**Comment by orta** [comment]
Yeah, true that, I found the template: https://github.com/danger/danger/blob/132c5d1a415656bee49d451b861569a6a08cfc30/lib/danger/comment_generators/bitbucket_server.md.erb#L1-L20

**Comment by azz** [comment]
(not sure what happened with that merge commit, will have to roll back)

**Comment by orta** [comment]
The above message is an integration test that should only sending its messages to STDOUT btw, if you're wondering wha could have triggered it

File: source/platforms/bitbucket_server/BitBucketServerGit.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    deleted_files,
+    commits,
+  }
+}
+
+const bitBucketServerDiffToGitStructuredDiff = (diffs: BitBucketServerDiff[]): GitStructuredDiff => {
+  return diffs.map(diff => ({
+    from: diff.source && diff.source.toString,
+    to: diff.destination && diff.destination.toString,
+    chunks: diff.hunks.map(hunk => ({
```
Comment (azz) [inline source/platforms/bitbucket_server/BitBucketServerGit.ts]: I have no idea what the difference between a hunk and a chunk is...

Comment (orta) [reply source/platforms/bitbucket_server/BitBucketServerGit.ts]: All the piece need for a male model catalogue

**Comment by azz** [comment]
I think all the functionality is done. I'll start writing tests tomorrow and look at documentation.

**Comment by azz** [comment]
Added tests. 🔬

**Comment by azz** [comment]
@orta know what's up with this build?
image

🔗  https://travis-ci.org/danger/danger-js/jobs/345868483

**Comment by orta** [comment]
The TypeScript doesn't seem to compile, https://travis-ci.org/danger/danger-js/jobs/345868483#L593-L596

```
source/ci_source/providers/Nevercode.ts:46:15 - error TS6133: 'branchName' is declared but its value is never read.
```

Which cascades through some of the next tests.

**Comment by orta** [comment]
Wow @azz, this is so much work 👍

**Comment by orta** [comment]
I've got it green, and extended the DSL with some JSDocs

**Comment by azz** [comment]
Awesome, I'll try and find some time in the next few days to write docs and give it a test on a project at work.

**Comment by orta** [comment]
Cool cool, once you're happy with the docs in here I can go take it to the more gnarly [getting started page](https://gitlab.com/danger-systems/danger.systems/blob/master/static/source/js/guides/getting_started.html.slim)

**Comment by azz** [comment]
Had a busy week at work and didn't get a chance to try it out! I'll try again early this coming week.

**Comment by orta** [comment]
OK @azz - I've gone and done a docs run, and prepared for it on the website too ( https://gitlab.com/danger-systems/danger.systems/merge_requests/168 )

Want to give me a lint for it all and say whether this is good to go?

**Comment by zdenektopic** [comment]
For some reason when installing from your branch I am not able to access danger command using yarn.

**Comment by orta** [comment]
It's probably not being compiled by TypeScript

**Comment by zdenektopic** [comment]
So how would I fix it? I am not JS dev, but really would like to adopt the JS version of danger with BBS support finally :)

**Comment by orta** [comment]
I'd wait TBH if you're not a JS dev, it shouldn't be too long

**Comment by azz** [comment]
:+1: :+1: :+1:

**Comment by orta** [comment]
ok, let's try get this rolling

**Comment by langovoi** [comment]
@azz i found that API `/diff` don't return diff for files in sub-sub-directory from root.

So `modified_files`, `deleted_files` and `creates_files` are wrong.

Do you have this problem with your Bitbucket Server? I could not find any information or issues about this bug/behavior.

**Pull request merged**